### PR TITLE
Interpret BERT Models Part 1

### PR DESCRIPTION
**Pull request by NarineK**
This PR shows how to use interpretation hooks to examine and better understand embeddings, sub-embeddings, BERT, and attention layers for question answering task fine tuned on SQUAD dataset.

### CODE REVIEW

File: tutorials/Bert_SQUAD_Interpret.ipynb [hunk 1]
```diff
@@ -0,0 +1,1111 @@
...
+   "metadata": {},
+   "source": [
+    "The first step is to fine tune BERT model on SQUAD dataset. This can be easiy accomplished by following the steps described in hugging face's official web site: https://github.com/huggingface/transformers#run_squadpy-fine-tuning-on-squad-for-question-answering "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "After we pretrained the model, we can loading the tokenizer and pre-trained BERT model using commands described below."
```
Comment (orionr) [inline tutorials/Bert_SQUAD_Interpret.ipynb]: Nit: `pretrained the model, we can loading` -> `pretrain the model, we can load`

File: tutorials/Bert_SQUAD_Interpret.ipynb [hunk 2]
```diff
@@ -0,0 +1,1111 @@
...
+   "outputs": [],
+   "source": [
+    "device = torch.device(\"cuda:0\" if torch.cuda.is_available() else \"cpu\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The first step is to fine tune BERT model on SQUAD dataset. This can be easiy accomplished by following the steps described in hugging face's official web site: https://github.com/huggingface/transformers#run_squadpy-fine-tuning-on-squad-for-question-answering "
```
Comment (orionr) [inline tutorials/Bert_SQUAD_Interpret.ipynb]: Nit: `fine tune` ->  `fine-tune`

File: tutorials/Bert_SQUAD_Interpret.ipynb [hunk 3]
```diff
@@ -0,0 +1,1111 @@
...
+    "    return pred.max(1).values"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's interpret parent `BertEmbeddings` layer. For that reason we separate embedding layer and precompute the embeddings. `configure_interpretable_embedding_layer` function helps us to do that. It returns `InterpretableEmbeddingBase` layer that wraps `BertEmbeddings` and can be used to access the embedding vectors. Note that we need to remove `InterpretableEmbeddingBase` wrapper from our model using `remove_interpretable_embedding_layer` function after we finish interpretation.\n",
+    "\n",
+    "Note that there is a way to interpret BertEmbeddings without configuring interpretable embedding layers that we will make available soon, However it is important to not that configuring interpretable embedding layers gives us more flexibility."
```
Comment (orionr) [inline tutorials/Bert_SQUAD_Interpret.ipynb]: Nit: `it is important to not that configuring` -> `it is important to note that configuring`

File: tutorials/Bert_SQUAD_Interpret.ipynb [hunk 4]
```diff
@@ -0,0 +1,1111 @@
...
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/home/narine/captum/captum/attr/_models/base.py:181: UserWarning: In order to make embedding layers more interpretable they will\n",
+      "        be replaced with an interpretable embedding layer which wraps the\n",
+      "        original embedding layer and takes word embedding vectors as inputs of\n",
+      "        the forward function. This allows to generate baselines for word\n",
```
Comment (orionr) [inline tutorials/Bert_SQUAD_Interpret.ipynb]: Nit: `This allows to generate` -> `This allows us to generate`
and might be in the source code?

File: tutorials/Bert_SQUAD_Interpret.ipynb [hunk 5]
```diff
@@ -0,0 +1,1111 @@
...
+      "        embeddings and compute attributions for each embedding dimension.\n",
+      "        The original embedding layer must be set\n",
+      "        back by calling `remove_interpretable_embedding_layer` function\n",
+      "        after model interpretation is finished.\n",
+      "  after model interpretation is finished.\"\"\"\n"
```
Comment (orionr) [inline tutorials/Bert_SQUAD_Interpret.ipynb]: Remove this in the source code?

File: tutorials/Bert_SQUAD_Interpret.ipynb [hunk 6]
```diff
@@ -0,0 +1,1111 @@
...
+    "ref_token_id = tokenizer.encode(ref)\n",
+    "sep_token_id = tokenizer.encode(sep)\n",
+    "cls_token_id = tokenizer.encode(cls)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Below we define a set of helper function fo constructing refenreces / baselines for word tokens, token types and position ids. We also provide separate helper functions that allow to construct the sub-embeddings and corresponding refenrences for all subembeddings of `BertEmbeddings` layer."
```
Comment (orionr) [inline tutorials/Bert_SQUAD_Interpret.ipynb]: Nit: `fo constructing refenreces` -> `for constructing references`

Comment (orionr) [reply tutorials/Bert_SQUAD_Interpret.ipynb]: Nit: `corresponding refenrences for all subembeddings` -> `corresponding references for all sub-embeddings`

File: tutorials/Bert_SQUAD_Interpret.ipynb [hunk 7]
```diff
@@ -0,0 +1,1111 @@
...
+    "question, text = \"What is important to us?\", \"It is important to us to include, empower and support humans of all kinds.\"\n",
+    "\n",
+    "ground_truth = 'to include, empower and support humans of all kinds'"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's numericalize the question, the input text and generate corresponding references for all three sub embedding (word, token type and position embeddings) types using our helper functions defined above."
```
Comment (orionr) [inline tutorials/Bert_SQUAD_Interpret.ipynb]: Nit: `sub embedding` -> `sub-embedding`

File: tutorials/Bert_SQUAD_Interpret.ipynb [hunk 8]
```diff
@@ -0,0 +1,1111 @@
...
...eqBZ/udwb9LqLPmCuTJ0+2Tcn7wbv6kp81a5ZbxQ7z/b0iaSdMIIAAAggggAACCCCAAAIIIIAAAggggAACJRYgeC8xcDF3n28gpP5b/dBdf9Tef//9zeabb27UF/uZZ55pvv76a3uq6YJ31VpXX+xh0fqun3e37SWXXGJuuOGGaNXu3bubhg0bRtPhiB5C1a4jeA9lymua57O87hdnGy/AcxzvwtzyEbjxxhvNxRdfHJ3wrbfeav76179G025EAbx+B/DLV199ZZum95uMV7/ujRo18lczv/32m6mqqrLz6tWrZ1ZYYQXbYo77fUBBurq5CUvPnj3NAw88EM1W3+9qEccP3uNqy+f7e0V0IEYQQAABBBBAAAEEEEAAAQQQQAABBBBAAIEaECB4rwHkYh0in0BIfxhv2rRpVAtNf9AeNWqUWW211aLT8mupufBcC/0/hGtaf6T3+2dVWN+qVSstsqVr167m3nvvNQ8//LA5+eST3Wwb2Cu498vPP/9s1P+siv7ovsYaaxC8+0BlOM7zWYY3jVNOEeA5TiFhRpkJqMuXQw45JDrrvffe2wwdOtQoIPfLkiVLTLNmzaLfD7RMQfjgwYPNOeecE62qbffZZ59oWrXk9buEqy2v/tjVyk1YU16/I/i/a+h4W2+9ddTUvR+w+79vuJr30QETI/n+XuHvg3EEEEAAAQQQQAABBBBAAAEEEEAAAQQQQKDUAgTvpRYu4v7DQGivvfYy+oN3uqJa5h07djQtWrSIVjn66KPNXXfdFU2HtcwzBe+quX777beb5Zdf3vz+++82XPeblO/bt6/9Y/3nn39u2rRpEx3j0EMPNYMGDYpqzKl/+R133DH6o73rVzY8F11vjx49ov0wUrsFeD5r9/3h7HIT4DnOzYm1aq/A7Nmzjbpw8cspp5xiLrjgArPWWmvZ2QrN+/XrZ1Q73hVXS/3DDz807du3d7PtS3gjRoyIfoYPGDDA9O7dO1qun+8K3dWyjlrYcUXHvP7666PA/84770wK9NV8/E033WRXzxa85/t7hTsXhggggAACCCCAAAIIIIAAAggggAACCCCAQE0IELzXhHKRjhEGQtl2qz7Xv/jii6Qa79pGf2hv3bq1mTBhgunTp08UgGtZuhpoWqaiYH6nnXYy7777rhk7duwfMxP/VfP1n376qa25rpkK0/XHeL8ogF+wYI
```
Comment (orionr) [inline tutorials/Bert_SQUAD_Interpret.ipynb]: Nit: `subembeddings` ->  `sub-embeddings`

Comment (orionr) [inline tutorials/Bert_SQUAD_Interpret.ipynb]: Nit: `a sepcific token` -> `a specific token`

Comment (orionr) [inline tutorials/Bert_SQUAD_Interpret.ipynb]: Nit: `Please, be patient!` -> `Please be patient!`

Comment (orionr) [inline tutorials/Bert_SQUAD_Interpret.ipynb]: Nit: `interqurtile` ->  `interquartile`

File: tutorials/Bert_SQUAD_Interpret.ipynb (hunk 5 above)
Comment (NarineK) [reply tutorials/Bert_SQUAD_Interpret.ipynb]: Interestingly there is only one encounter of `after model interpretation is finished` in the code

**Pull request closed without merging**